File: app/src/main/java/com/sample/framework/mvvm/java/base/ViewModelFactory.java
package com.sample.framework.mvvm.java.base;

import androidx.annotation.NonNull;
import androidx.core.util.Supplier;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class ViewModelFactory<T extends BaseViewModel> implements ViewModelProvider.Factory {

    private final Class<T> viewModelClass;
    private final Supplier<T> viewModelSupplier;

    public ViewModelFactory(Class<T> viewModelClass, Supplier<T> viewModelSupplier) {
        this.viewModelClass = viewModelClass;
        this.viewModelSupplier = viewModelSupplier;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(viewModelClass)) {
            return (T) viewModelSupplier.get();
        } else {
            throw new IllegalArgumentException("Unknown class name");
        }
    }

}

File: app/src/main/java/com/sample/framework/mvvm/kotlin/ui/base/ViewModelFactory.kt
package com.sample.framework.mvvm.kotlin.ui.base

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.sample.framework.mvvm.kotlin.data.api.ApiHelper
import com.sample.framework.mvvm.kotlin.data.repository.MainRepository
import com.sample.framework.mvvm.kotlin.ui.main.viewmodel.MainViewModel

open class ViewModelFactory(private val apiHelper: ApiHelper) : ViewModelProvider.Factory {

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(MainViewModel::class.java)) {
            return MainViewModel(MainRepository(apiHelper)) as T
        }
        throw IllegalArgumentException("Unknown class name")
    }

}